### DISCUSSION TITLE
TS Parser failed

### DISCUSSION THREAD

**Discussion by solaeus** [helix-editor/helix#9270] [Troubleshooting]
### Summary

The editor frequently loses syntax highlighting for the active buffer. Other buffers still have syntax highlighting and closing and reopening the buffer restores syntax highlighting.

### Reproduction Steps

This has been happening while editing Rust in a modestly-sized codebase but I have also experienced the issue with a JavaScript file.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2024-01-06T05:30:47.077 helix_core::syntax [ERROR] TS parser failed, disabling TS for the current buffer: Err(Cancelled)
2024-01-06T05:36:55.508 helix_view::editor [ERROR] editor error: No code actions available
2024-01-06T05:38:46.810 helix_lsp::transport [ERROR] Tried sending response into a closed channel (id=Num(5953)), original request likely timed out
2024-01-06T05:49:02.078 helix_view::document [ERROR] TS parser failed, disabling TS for the current buffer: Err(Cancelled)
```

</details>

### Platform

Linux

### Terminal Emulator

kitty 0.31.0

### Installation Method

dnf

### Helix Version

helix 23.10

**Comment by kirawi** [comment]
Can you take an asciicinema cast; the description is unclear what you mean.

**Reply by solaeus** [reply]
I don't understand the request for a video of the editor showing no syntax highlighting. It's a screen with white text.

**Comment by the-mikedavis** [comment]
From the log file it looks like we're giving up on parsing the files because we've hit an internal time limit. This is by design to make sure we don't spend too much work waiting on tree-sitter for especially large files. Can you link to the files that caused this?

**Comment by solaeus** [comment]
This now happened in multiple repositories and with multiple languages. It's happened while looking through the Rust standard library and while viewing small, single files by themselves. It has even happened while editing a README so it is not associated with the size of the file. If you really want to see the repository it is my only pinned repo. This is where most of my encounters with this error happen but that's only because it's where I've been working. Helix is dropping its TS highlighting mid-edit on all sorts of documents. If other users experience this, they will not find this bug report in the issues because it was moved to discussions.

**Reply by kirawi** [reply]
It would be helpful if you can take a capture of what you're experiencing and offer reproduction steps. This issue is unactionable otherwise.

Edit: Ah, I see that you replied to my previous message.

Tree-sitter is disabled if it takes more than 500ms to parse the syntax: https://github.com/helix-editor/helix/pull/7028

**Comment by k9withabone** [comment]
I am also experiencing this fairly frequently. It only started happening after I switched from [the copr](https://copr.fedorainfracloud.org/coprs/varlad/helix/) to the official Fedora package a few weeks ago. Perhaps it has something to do with how Fedora is building it?

**Reply by bu6n** [reply]
I have the same problem on Fedora 39, also using the official package. My machine is a M1 Macbook Pro, so I don't think that the tree-sitter parsing would take long. It also happens across different languages for me. Did you find a workaround? Or is there at least a command to re-parse without restarting the editor? I'm currently working on a file that's only 1200 lines and it literally happens every 5 minutes.

**Reply by k9withabone** [reply]
> Did you find a workaround? Or is there at least a command to re-parse without restarting the editor?

I don't have to completely restart the editor. I just have to close and reopen the buffer, which means you lose your place and editing history, but at least then you don't have to wait for your language server to start again.

**Reply by bu6n** [reply]
Yes that's true. What hardware do you have?

**Reply by k9withabone** [reply]
Surface Pro 7, kernel 6.7.2-1.surface.fc39.x86_64

I don't think that this has anything to do with hardware differences. It's probably just something weird with the packaging.

**Reply by bu6n** [reply]
The package source is here: https://src.fedoraproject.org/rpms/helix/blob/f39/f/helix.spec
I don't see anything special there.

**Reply by k9withabone** [reply]
I don't know enough about packaging to be able to compare the two but the source for the copr is here: https://copr-dist-git.fedorainfracloud.org/cgit/varlad/helix/helix.git/tree/helix.spec

**Reply by bu6n** [reply]
@k9withabone, I tried to compile from the code directly and I have not gotten the error so far. It seems that the Fedora 39 package uses pre-compiled grammars ([link](https://src.fedoraproject.org/rpms/helix/blob/f39/f/helix.spec#_34)), whereas the copr package rebuilds everything ([link](https://copr-dist-git.fedorainfracloud.org/cgit/varlad/helix/helix.git/tree/helix.spec#n26)). I'm not sure what's going on, but I hope this helps you somehow? I only needed the packages `rust`, `cargo` and `gcc-c++`.

**Comment by ikasoba** [comment]
Oddly enough, in my environment, I was able to get around this by creating a `.git` directory under the current directory.
